src/content/components/TimelineCanvas.js

```diff
@@ -0,0 +1,176 @@
+  containerHeight: CssPixels,
+  className: string,
+  onDoubleClickItem: HoveredItem => void,
+  getHoveredItemInfo: HoveredItem => string,
+  drawCanvas: (CanvasRenderingContext2D, HoveredItem) => void,
+  hitTest: (x: CssPixels, y: CssPixels) => null | HoveredItem,
+};
+
+require('./TimelineCanvas.css');
+
+export default class TimelineCanvas extends Component {
+
+  props: Props
+  _requestedAnimationFrame: boolean
+  _devicePixelRatio: 1
+  _ctx: CanvasRenderingContext2D
+  state: {
+    hoveredItem: null | HoveredItem;
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this._requestedAnimationFrame = false;
+    this._devicePixelRatio = 1;
+    this.state = { hoveredItem: null };
+
+    (this: any).onMouseMove = this.onMouseMove.bind(this);
+    (this: any).onMouseOut = this.onMouseOut.bind(this);
+    (this: any).onDoubleClick = this.onDoubleClick.bind(this);
+    (this: any).getHoveredItemInfo = this.getHoveredItemInfo.bind(this);
+  }
+
+  shouldComponentUpdate() {
+    // If the parent updates, always re-render.
+    return true;
+  }
+
+  _scheduleDraw() {
+    const { className, drawCanvas } = this.props;
+    if (!this._requestedAnimationFrame) {
+      this._requestedAnimationFrame = true;
+      window.requestAnimationFrame(() => {
+        this._requestedAnimationFrame = false;
+        if (this.refs.canvas) {
+          timeCode(`${className} render`, () => {
+            this._prepCanvas();
+            drawCanvas(this._ctx, this.state.hoveredItem);
+          });
+        }
+      });
+    }
+  }
+
+  _prepCanvas() {
+    const {canvas} = this.refs;
+    const {containerWidth, containerHeight} = this.props;
+    const {devicePixelRatio} = window;
+    const pixelWidth: DevicePixels = containerWidth * devicePixelRatio;
+    const pixelHeight: DevicePixels = containerHeight * devicePixelRatio;
+    if (!this._ctx) {
+      this._ctx = canvas.getContext('2d');
+    }
+    if (canvas.width !== pixelWidth || canvas.height !== pixelHeight) {
+      canvas.width = pixelWidth;
+      canvas.height = pixelHeight;
+      canvas.style.width = containerWidth + 'px';
+      canvas.style.height = containerHeight + 'px';
+      this._ctx.scale(this._devicePixelRatio, this._devicePixelRatio);
+    }
+    if (this._devicePixelRatio !== devicePixelRatio) {
+      // Make sure and multiply by the inverse of the previous ratio, as the scaling
+      // operates off of the previous set scale.
+      const scale = (1 / this._devicePixelRatio) * devicePixelRatio;
+      this._ctx.scale(scale, scale);
+      this._devicePixelRatio = devicePixelRatio;
+    }
+    return this._ctx;
+  }
+
+  onMouseMove(event: SyntheticMouseEvent) {
+    const { canvas } = this.refs;
+    if (!canvas) {
+      return;
+    }
+
+    const rect = canvas.getBoundingClientRect();
+    const x: CssPixels = event.pageX - rect.left;
+    const y: CssPixels = event.pageY - rect.top;
+
+    const maybeHoveredItem = this.props.hitTest(x, y);
```

Comment:
I think this is the only way to to have it be fully reactive to user input, otherwise the UI feels sluggish. Here's the actual perf on my machine:

56ms for a noop mouse move.
<img width="332" alt="image" src="https://cloud.githubusercontent.com/assets/1588648/25815273/b19477d6-33e5-11e7-8a8f-a9b185db3969.png">

14ms for changing the hover target:

<img width="872" alt="image" src="https://cloud.githubusercontent.com/assets/1588648/25815429/f33bcb62-33e5-11e7-8c1d-9cfa063808cb.png">

I would prefer to keep the UI responsive, but if we care about the performance, we could only redraw the hovered frames. Let me know if we should file a follow-up bug for that.